include/xtensor/xbuilder.hpp

```diff
@@ -708,107 +708,115 @@ namespace xt
     }
 
     /**
-     * @function diagonal(const xexpression<T>& arr)
      * @brief Returns the elements on the diagonal of arr
      *
      * @param arr the input array
+     * @returns xexpression with values of the diagonal
      *
-     * @return xexpression with values of the diagonal
+     * \code{.cpp}
+     * xt::xarray<double> a = {{1, 2, 3},
+     *                         {4, 5, 6}
+     *                         {7, 8, 9}};
+     * auto b = xt::diagonal(a); // => {1, 5, 9}
+     * \endcode
      */
     template <class E>
     inline auto diagonal(E&& arr)
     {
         using CT = xclosure_t<E>;
+        std::vector<std::size_t> shape = {arr.shape()[0]};
```

Comment:
you should be able to only store a single integer. This would prevent a heap allocation.

```
auto s = arr.shape()[0];
```

then pass `{ s }` to `make_xgenerator`.